### PR TITLE
CC-12856 Introduce the Pre Checkout Logic

### PR DESCRIPTION
**Pull request by AsonUnique**
#### Overview:
- Jira ticket: https://spryker.atlassian.net/browse/CC-12856

### CODE REVIEW

**Review by romanhavrylko — APPROVED** [review]

File: composer.json [hunk 1]
```diff
@@ -32,6 +34,13 @@
       "SprykerEcoTest\\": "tests/SprykerEcoTest/"
     }
   },
+  "suggest": {
+    "spryker-eco/avalara-tax-product-option": "If you want to calculate product option taxes using Avalara",
```
Comment (abitskil) [inline composer.json]: * Add the dot to the end, check the sentence below.

Comment (abitskil) [inline composer.json]: * If you want to calculate product option taxes using Avalara tax provider.

File: src/SprykerEco/Zed/AvalaraTax/Business/Checker/AvalaraTaxQuoteChecker.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+    /**
+     * @param \Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     * @param \Generated\Shared\Transfer\CheckoutResponseTransfer $checkoutResponseTransfer
+     *
+     * @return bool
+     */
+    public function isQuoteTaxCalculationValid(QuoteTransfer $quoteTransfer, CheckoutResponseTransfer $checkoutResponseTransfer): bool
+    {
+        $quoteTransfer = $this->calculationFacade->recalculateQuote($quoteTransfer);
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/Business/Checker/AvalaraTaxQuoteChecker.php]: * I think we don't need to recalculate a quote here.
* After adding an item to the cart, recalculation is always called.

File: src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php [hunk 3]
```diff
@@ -99,7 +99,6 @@ public function executeAvalaraCreateTransaction(
             $avalaraApiLogTransfer
                 ->setIsSuccessful(true)
                 ->setResponse($this->utilEncodingService->encodeJson((array)$transactionModel));
-            file_put_contents(APPLICATION_ROOT_DIR . '/response-5.json', json_encode($transactionModel, JSON_PRETTY_PRINT));
         } catch (Throwable $e) {
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php]: * Add the caching to the request.

File: composer.json (hunk 1 above)
Comment (abitskil) [reply composer.json]: Resolved.

Comment (abitskil) [reply composer.json]: Resolved.

File: src/SprykerEco/Zed/AvalaraTax/Business/Checker/AvalaraTaxQuoteChecker.php (hunk 2 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/Business/Checker/AvalaraTaxQuoteChecker.php]: Resolved.

File: src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php [hunk 4]
```diff
@@ -78,6 +84,10 @@ public function executeAvalaraCreateTransaction(
         CalculableObjectTransfer $calculableObjectTransfer,
         string $transactionTypeId
     ): AvalaraCreateTransactionResponseTransfer {
+        if (static::$avalaraCreateTransactionResponseCache) {
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php]: * The cache should also take into account changes in the `CalculableObjectTransfer`.

File: src/SprykerEco/Zed/AvalaraTax/Communication/Plugin/Checkout/AvalaraTaxCheckoutPreConditionPlugin.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * @method \SprykerEco\Zed\AvalaraTax\AvalaraTaxConfig getConfig()
+ * @method \SprykerEco\Zed\AvalaraTax\Business\AvalaraTaxFacadeInterface getFacade()
+ */
+class AvalaraTaxCheckoutPreConditionPlugin extends AbstractPlugin implements CheckoutPreConditionPluginInterface
+{
+    /**
+     * {@inheritDoc}
+     * - Recalculates quote.
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/Communication/Plugin/Checkout/AvalaraTaxCheckoutPreConditionPlugin.php]: * Remove the first sentence.

File: src/SprykerEco/Zed/AvalaraTax/Business/AvalaraTaxFacadeInterface.php [hunk 6]
```diff
@@ -57,4 +59,20 @@ public function expandProductConcreteWithAvalaraTaxCode(ProductConcreteTransfer
      * @return \Generated\Shared\Transfer\CartChangeTransfer
      */
     public function expandCartItemsWithAvalaraTaxCode(CartChangeTransfer $cartChangeTransfer): CartChangeTransfer;
+
+    /**
+     * Specification:
+     * - Recalculates quote.
```
Comment (abitskil) [inline src/SprykerEco/Zed/AvalaraTax/Business/AvalaraTaxFacadeInterface.php]: * Remove the first sentence.

File: tests/SprykerEcoTest/Zed/AvalaraTax/Business/AvalaraTaxFacadeTest.php [hunk 7]
```diff
@@ -203,6 +262,19 @@ protected function getTransactionModelMock(string $fileName): stdClass
         return json_decode($avalaraApiResponseJson);
     }
 
+    /**
+     * @param \PHPUnit\Framework\MockObject\MockObject|\Generated\Shared\Transfer\QuoteTransfer $quoteTransfer
+     *
+     * @return \SprykerEco\Zed\AvalaraTax\Dependency\Facade\AvalaraTaxToCalculationFacadeInterface
+     */
+    protected function createCalculationFacadeMock(QuoteTransfer $quoteTransfer): AvalaraTaxToCalculationFacadeInterface
```
Comment (abitskil) [inline tests/SprykerEcoTest/Zed/AvalaraTax/Business/AvalaraTaxFacadeTest.php]: * This mock is not needed anymore.
* Remove it from places where it is used.

File: src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php (hunk 3 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php]: Resolved.

File: src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php (hunk 4 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/Business/Executor/AvalaraTransactionExecutor.php]: Resolved.

File: src/SprykerEco/Zed/AvalaraTax/Communication/Plugin/Checkout/AvalaraTaxCheckoutPreConditionPlugin.php (hunk 5 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/Communication/Plugin/Checkout/AvalaraTaxCheckoutPreConditionPlugin.php]: Resolve.

File: src/SprykerEco/Zed/AvalaraTax/Business/AvalaraTaxFacadeInterface.php (hunk 6 above)
Comment (abitskil) [reply src/SprykerEco/Zed/AvalaraTax/Business/AvalaraTaxFacadeInterface.php]: Resolve.

File: tests/SprykerEcoTest/Zed/AvalaraTax/Business/AvalaraTaxFacadeTest.php (hunk 7 above)
Comment (abitskil) [reply tests/SprykerEcoTest/Zed/AvalaraTax/Business/AvalaraTaxFacadeTest.php]: Resolved.

**Review by abitskil — APPROVED** [review]

**Pull request merged**